libavpipe/src/avpipe_mux.c

```diff
@@ -424,42 +434,84 @@ avpipe_mux(
     }
 
     for (int i=0; i<MAX_STREAMS; i++) {
-        first_pts_array[i] = AV_NOPTS_VALUE;
+        last_pts_array[i] = AV_NOPTS_VALUE;
+        pts_per_frame[i] = AV_NOPTS_VALUE;
+        next_pts_array[i] = 0;
     }
 
     pkts = xctx->pkt_array;
     valid_pkts = xctx->is_pkt_valid;
     in_mux_ctx = xctx->in_mux_ctx;
+    int stream_count = in_mux_ctx->audio_count + in_mux_ctx->caption_count + 1;
+
+    // Set pts_per_frame for streams that have frame rate info
+    for (int i=0; i < stream_count; i++) {
+        AVRational avg_frame_rate = xctx->in_muxer_ctx[i].format_context->streams[0]->avg_frame_rate;
+        AVRational time_base = xctx->in_muxer_ctx[i].format_context->streams[0]->time_base;
+        if (avg_frame_rate.num == 0 || time_base.num == 0) {
+            elv_dbg("avpipe_mux stream %d avg_frame_rate or time_base is 0", i);
+            continue;
+        }
+        pts_per_frame[i] = av_rescale_q(1, av_inv_q(time_base), avg_frame_rate);
+        elv_warn("avpipe_mux stream %d avg_frame_rate %d/%d tb %d/%d, pts per frame: %d", i, avg_frame_rate.num, avg_frame_rate.den, time_base.num, time_base.den, pts_per_frame[i]);
+    }
+
+    for (int i=0; i < stream_count; i++) {
 
-    for (int i=0; i<in_mux_ctx->last_caption_index + in_mux_ctx->last_audio_index + 1; i++) {
         ret = av_read_frame(xctx->in_muxer_ctx[i].format_context, &pkts[i]);
         if (ret >= 0) {
             valid_pkts[i] = 1;
-            first_pts_array[i] = pkts[i].pts;
+            last_pts_array[i] = pkts[i].pts;
         }
     }
 
     while (1) {
         ret = get_next_packet(xctx, &pkt);
         if (ret < 0)
             break;
+        stream_index = ret;
+
+        if (last_pts_array[stream_index] != AV_NOPTS_VALUE && found_keyframe) {
+            if (pts_per_frame[stream_index] == AV_NOPTS_VALUE) {
+                // set pts per frame, because this is the second packet we have seen for this stream
+                pts_per_frame[stream_index] = pkt.pts - last_pts_array[stream_index];
+            } else {
+                int64_t pts_delta = pkt.pts - last_pts_array[stream_index];
+                // Check if the pts delta is more than 20% away from what we expected
+                if (labs(pts_delta - pts_per_frame[stream_index]) > (pts_per_frame[stream_index] / 5)) {
+                    elv_dbg("avpipe_mux stream %d pts_delta=%"PRId64", expected=%"PRId64", pts=%"PRId64", last_pts=%"PRId64"",
```

Comment:
I think we should elv_dbg if the delta is different at all.  We don't expect that many imperfections (for example most I have seen is 3 or 4 per minute)

Reply:
Need to be careful not to bog down the server if somehow it gets into a bad state where it's logging every frame